### PR TITLE
feat: Return logits in PyTorch flow

### PR DESCRIPTION
**Pull request by tongyuantongyu**

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 1]
```diff
@@ -35,10 +35,197 @@
...
+        self._py_result = py_result
+
+    def __getstate__(self):
+        return self._response, self._py_result
+
+    def __setstate__(self, state):
+        self._response, self._py_result = state
+
+    @property
+    def result(self) -> tensorrt_llm.bindings.executor.Result:
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:200]: The return type hint should be `LlmResult`?

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 2]
```diff
@@ -35,10 +35,197 @@
...
+
+class LlmResponse:
+    """LlmResponse wraps `bindings.executor.Response` but detour some features to Python implementation"""
+
+    def __init__(self, response: tensorrt_llm.bindings.executor.Response,
+                 py_result: PyResult):
+        self._response = response
+        self._py_result = py_result
+
+    def __getstate__(self):
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:193]: Why we need to override `__getstate__` and `__setstate__` here?

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 3]
```diff
@@ -49,6 +236,27 @@ def __init__(self, *args, **kwargs):
...
+
+        self.py_result = PyResult(self.py_prompt_len, self.py_max_new_tokens,
+                                  return_logits_device_memory, self.streaming,
+                                  return_log_probs, return_context_logits,
+                                  return_generation_logits)
+
+    def create_response(
+            self,
+            use_fast_logits=False,
+            mpi_world_rank=0) -> tensorrt_llm.bindings.executor.Response | None:
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:263]: The return type hint should be `LlmResponse | None`?

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 4]
```diff
@@ -20,40 +20,53 @@
...
-    def update_requests(self, scheduled_requests: ScheduledRequests,
-                        new_tensors_host: Dict[str, torch.tensor],
-                        decoder_event: torch.cuda.Event):
+    def update_requests(self, decoder_state: DecoderState) -> None:
         raise NotImplementedError
 
 
 class DummyDecoder(Decoder):
 
-    def setup_decoder(self, scheduled_requests: ScheduledRequests,
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: I think we can remove `DummyDecoder` class directly, it's useless.

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 5]
```diff
@@ -20,40 +20,53 @@
...
 from tensorrt_llm.mapping import Mapping
 
-from .llm_request import LlmRequest, LlmRequestState
+from .llm_request import LlmRequest, LlmRequestState, LogProbs
 from .scheduler import ScheduledRequests
 
 
+@dataclass
+class DecoderState:
+    scheduled_requests: ScheduledRequests
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: I notice that we introduce `DecoderState` and `BatchDecoder` in this PR. It makes sense to me to use a data class to encapsulate some states. 

Just for better code review, could you please split this PR to some small PRs:

1. a PR for introducing `DecoderState` and `BatchDecoder` only
2. a PR for support return logits based on PR 1.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 1 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:200]: `LlmResult` is masquerading `tensorrt_llm.bindings.executor.Result`. Since `tensorrt_llm.bindings.executor.Result` is not constructed by `LlmResponse` we can't make `LlmResult` inherit `tensorrt_llm.bindings.executor.Result` unless we are fine with reassigning `__class__` (I recently realized modelopt is doing this btw https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/opt/dynamic.py#L657).

I may give it a try to see if that works and doesn't break anything.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 2 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:193]: We need to pass `LlmResponse` between different processes, therefore it must be pickleable.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 3 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:263]: Same as above.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 5 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: Created https://github.com/NVIDIA/TensorRT-LLM/pull/3234.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 1 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:200]: Pybind doesn't like us doing that: `__class__ assignment: 'LlmResponse' deallocator differs from 'tensorrt_llm.bindings.executor.Response'`. So I guess we have to "lie" in the return type hint.

**Comment by hchings** [comment]
Hi @tongyuantongyu , I'm also adding logprobs and prompt logprobs for LLMAPI here: https://github.com/NVIDIA/TensorRT-LLM/pull/3388 per discussion w/ @Superjomn. 

We need to consolidate the input & output API (the one we're doing is aligned w 3rd party FW), and where to compute logprobs. Originally we were thinking to calculate in executor layer for both TRT & PyT backends, but open to other better ideas.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 6]
```diff
@@ -35,10 +35,199 @@
 }
 
 
+class LogitsStorage:
+    beam_width: int = -1
```
Comment (Superjomn) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py]: IIUC, are these two members ClassVar[int]?

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 7]
```diff
@@ -35,10 +35,199 @@
...
+    vocab_size: int
+
+    def __init__(self, seq_length: int, device_memory=True):
```
Comment (Superjomn) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py]: How about `use_device_memory` or something else? `device_memory` as a bool is a bit odd.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 8]
```diff
@@ -35,10 +35,199 @@
...
+        self._result = result
+        self._py_result = py_result
+
+    def __getattr__(self, item):
+        if item in self.py_result_properties:
+            return getattr(self._py_result, item)
+        return getattr(self._result, item)
+
+
+class LlmResponse:
```
Comment (Superjomn) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:185]: Maybe a minor issue, it is a bit odd to have `LlmResponse` in a file called `llm_request.py`, I thought it is for the `LlmRequest`. Should we partition the file or make the file name proper for a package of PyExecutor's input and output DS?

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 9]
```diff
@@ -35,10 +35,199 @@
...
+
+    def __init__(self, response: tensorrt_llm.bindings.executor.Response,
+                 py_result: PyResult):
+        self._response = response
+        self._py_result = py_result
+
+    def __getstate__(self):
+        return self._response, self._py_result
+
+    def __setstate__(self, state):
```
Comment (Superjomn) [inline tensorrt_llm/_torch/pyexecutor/llm_request.py:196]: Please add some tests for the pickle support of such core DS., We encountered some pickle failure that breaks the e2e tests and consumes more time to investigate.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 6 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py]: They are lazily initialized fields outside `__init__`, and these are type hints for them. I updated to setup placeholder values in `__init__`.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 8 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:185]: I don't have preference on this. @QiJune any opinion?

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 9 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:196]: I'm adding one in `unittest/_torch/test_return_logits.py`. Is there better location to add it?

**Comment by tongyuantongyu** [comment]
/bot run --only-multi-gpu-test --stage-list "DGX_H100-4_GPUs-PyTorch-2,DGX_H100-4_GPUs-CPP-1,DGX_H100-4_GPUs-PyTorch-1,DGX_H100-4_GPUs-TensorRT-2" --disable-fail-fast

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 9 above)
Comment (Superjomn) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:196]: I noticed that the `tests/unittest/_torch/pyexecutor` directory is missing. Typically, we add tests to a module with a similar path in the tests directory. Could we create a `tests/unittest/_torch/pyexecutor/test_llm_request.py` file to include the necessary tests? @QiJune do you have any concerns about this?

**Comment by suyoggupta** [comment]
@QiJune : could we merge this?

**Review by QiJune — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 8 above)
Comment (QiJune) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:185]: Yes, I think we can move it to `llm_response.py`. Let's do it in the following PR.

File: tensorrt_llm/_torch/pyexecutor/llm_request.py (hunk 9 above)
Comment (QiJune) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py:196]: @Superjomn  I agree with you, will refine it in the next PR.

**Comment by QiJune** [comment]
/bot skip --comment "skip a flaky test"
